### PR TITLE
refactor: create base collection class

### PR DESCRIPTION
**Pull request by rafilong**
towards #65

this PR is staged on top of #69 as refactor changes are cross-class; I encourage reviewing in sequential order

### CODE REVIEW

File: tests/unit/v1/test_collection.py [hunk 1]
```diff
@@ -32,10 +32,18 @@ def _make_one(self, *args, **kwargs):
 
     @staticmethod
     def _get_public_methods(klass):
-        return set(
-            name
-            for name, value in six.iteritems(klass.__dict__)
-            if (not name.startswith("_") and isinstance(value, types.FunctionType))
+        return set().union(
```
Comment (crwilcox) [inline tests/unit/v1/test_collection.py:35]: why this change?

Comment (rafilong) [reply tests/unit/v1/test_collection.py:35]: The previous test was testing the functions that exist in `CollectionReference`. Now that we've pulled some of those out to `BaseCollectionReference`, we need to test the public interface, which means including the functions that exist in the parent. This was my way of solving that problem.
